From: Sergio Costas Rodriguez <sergio.costas@canonical.com>
Date: Thu, 19 Jun 2025 13:42:23 +0200
Subject: a11y: Notify bad username

When the username is invalid, a text label is updated with the
problem. Unfortunately, the content is not notified to the user.

This patch checks for changes in the label, and uses
gtk_accessible_announce() to read out loud the text.

Forwarded: https://gitlab.gnome.org/GNOME/gnome-initial-setup/-/merge_requests/278
Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/+source/gnome-initial-setup/+bug/2116935
---
 gnome-initial-setup/pages/account/gis-account-page-local.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/gnome-initial-setup/pages/account/gis-account-page-local.c b/gnome-initial-setup/pages/account/gis-account-page-local.c
index 406b4be..8c2c2e8 100644
--- a/gnome-initial-setup/pages/account/gis-account-page-local.c
+++ b/gnome-initial-setup/pages/account/gis-account-page-local.c
@@ -305,7 +305,7 @@ validate (GisAccountPageLocal *page)
   GtkWidget *entry;
   const gchar *name, *username, *hostname;
   gboolean parental_controls_enabled;
-  gchar *tip;
+  g_autofree gchar *tip = NULL;
 
   g_clear_handle_id (&page->timeout_id, g_source_remove);
 
@@ -328,8 +328,10 @@ validate (GisAccountPageLocal *page)
   if (page->valid_username)
     set_entry_validation_checkmark (GTK_ENTRY (entry));
 
+  const gchar *current_label = gtk_label_get_text (GTK_LABEL (page->username_explanation));
+  if (!g_str_equal (tip, current_label) && !g_str_equal (current_label, ""))
+    gtk_accessible_announce (GTK_ACCESSIBLE (page), tip, GTK_ACCESSIBLE_ANNOUNCEMENT_PRIORITY_MEDIUM);
   gtk_label_set_text (GTK_LABEL (page->username_explanation), tip);
-  g_free (tip);
 
   begin_hostname_validation (page, hostname);
 
